pageserver/src/page_service.rs

```diff
@@ -326,13 +274,99 @@ impl PageServerHandler {
         broker_client: storage_broker::BrokerClientChannel,
         auth: Option<Arc<JwtAuth>>,
         connection_ctx: RequestContext,
+        cancel: CancellationToken,
     ) -> Self {
         PageServerHandler {
             _conf: conf,
             broker_client,
             auth,
             claims: None,
             connection_ctx,
+            cancel,
+        }
+    }
+
+    /// Wrap PostgresBackend::flush to respect our CancellationToken: it is important to use
```

Comment:
Would have preferred this to be implemented inside the postgres_backend crate, or, as a wrapper type around the `PostgresBackend` type, to ensure that it's compile-time-impossible to call the raw `pgb.flush()`.

Reply:
This is kind of a general concern around all our I/O interfaces: ideally nothing on the network should be called without a cancellation token, but equally the convention in rust is that async network libraries are cancel-safe and return std::io::Error or similar.

In this case I think we can do better in future, but fixing the postgres_backend interface in general is going to be a bit invasive because we an AsyncWrite implementor, and we also have the safekeeper that doesn't use cancellation token at all, hence an interface that currently uses shutdown_waiter futures instead of cancellation tokens.  It's bigger than the scope of this bugfix, for sure.